opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -102,51 +107,94 @@ public ExprType getExprType() {
    * @param mappingType A mapping type.
    * @return An instance or inheritor of `OpenSearchDataType`.
    */
-  public static OpenSearchDataType of(MappingType mappingType) {
+  public static OpenSearchDataType of(MappingType mappingType, Map<String, Object> innerMap) {
     var res = instances.getOrDefault(mappingType.toString(), null);
-    if (res != null) {
-      return res;
+    if (res == null) {
+      res = new OpenSearchDataType(mappingType);
     }
-    ExprCoreType exprCoreType = mappingType.getExprCoreType();
-    if (exprCoreType == ExprCoreType.UNKNOWN) {
-      switch (mappingType) {
+    switch (mappingType) {
+      case Object:
+      case Nested:
+        if (innerMap.isEmpty()) {
+          return res;
+        }
+        var im = innerMap.getOrDefault("properties", Map.of());
+        Map<String, OpenSearchDataType> properties =
+            parseMapping((Map<String, Object>) innerMap.getOrDefault("properties", Map.of()));
+        OpenSearchDataType objectDataType = res.cloneEmpty();
+        objectDataType.properties = properties;
+        return objectDataType;
+      case Text:
         // TODO update these 2 below #1038 https://github.com/opensearch-project/sql/issues/1038
-        case Text: return OpenSearchTextType.of();
-        case GeoPoint: return OpenSearchGeoPointType.of();
-        case Binary: return OpenSearchBinaryType.of();
-        case Ip: return OpenSearchIpType.of();
-        default:
-          throw new IllegalArgumentException(mappingType.toString());
-      }
+        Map<String, OpenSearchDataType> fields =
+            parseMapping((Map<String, Object>) innerMap.getOrDefault("fields", Map.of()));
+        return (fields != null) ? OpenSearchTextType.of(fields) : OpenSearchTextType.of();
+      case GeoPoint: return OpenSearchGeoPointType.of();
+      case Binary: return OpenSearchBinaryType.of();
+      case Ip: return OpenSearchIpType.of();
+      case Date:
+      case Time:
+      case Datetime:
+      case Timestamp:
+        return OpenSearchDateType.create(
+          (String) innerMap.getOrDefault("format", ""), mappingType);
+      default:
+        return res;
+    }
+  }
+
+  public static Map<String, OpenSearchDataType> parseMapping(Map<String, Object> indexMapping) {
+    Map<String, OpenSearchDataType> result = new LinkedHashMap<>();
+    if (indexMapping != null) {
+      indexMapping.forEach((k, v) -> {
+        var innerMap = (Map<String, Object>)v;
+        // by default, the type is treated as an Object if "type" is not provided
+        var type = ((String) innerMap
+            .getOrDefault(
+                "type",
+                "object"))
+            //TODO: Remove following line for 3.0 release. Temporary fix to prevent a breaking change.
+            .replace("date", "timestamp")
```

Comment:
You can remove this if you revert changes in `MappingType` enum

Reply:
Changes have been reverted.